gradle/libs.versions.toml

```
@@ -1,7 +1,9 @@
 [versions]
 kotlin = "1.8.22"
+foojay = "0.5.0"
 
 [libraries]
+foojay = { module = "org.gradle.toolchains:foojay-resolver", version.ref = "foojay" }
```

```suggestion
foojay = { module = "org.gradle.toolchains:foojay-resolver", version.ref = "0.5.0" }
```
This is also opnionated, but I personally like to use `[versions]` when there is a good reason for it.

---

If I move the `foojay-jvmToolchain` entry into the `[plugins]` section, then I would have to apply it via,

```
plugins {
    alias(libs.plugins.foojay.jvmToolchain)
}
```

However, `libs` doesn't seem to be available in `settings.gradle[.kts]` files, including in `settings` plugins. Investigating further lead me to this issue, https://github.com/gradle/gradle/issues/24876 – and according to that, the Gradle team has no plans to support `libs` ever being accessible in the `settings` context. Quote:

> This is the intended behavior, and there’s no plan to change it. This issue will be closed.
>
> Version catalogs can't be used before they have been evaluated, which happens during settings evaluation, which is AFTER evaluating ***settings plugins***. So basically a chicken and egg problem; nothing we can do about it.

(Emphasis mine.)

---

Although, in my opinion, the Gradle team could've at least supported the `libs` variable in precompiled script plugins of `settings` plugins. Actually, there's even another issue for that: “[Make version catalogs accessible from precompiled script plugins · Issue #15383 · gradle/gradle](https://github.com/gradle/gradle/issues/15383)” – this one is still open and not yet implemented.

---

You are right 🤦. You can't, it's an egg-chicken problem. Still, I don't feel comfortable doing all this refactoring just because of this. Thanks again for dedicating time, but giving the complexity the value added is just not worth it to me. If JDK 8 is too old that people simply don't install it anymore, I'd rather consider bumping it to JDK11/17.

---

Note also that, even if you bump it to JDK 11 or 17, and say I have JDK 19, the download logic of Gradle still needs to kick in, since Gradle wants exact versions of what you currently have installed.

---

Furthermore, you shouldn't bump the JVM bytecode output of your custom gradle plugin, since the output classes should stay simply in Java 8. I don't have a ticket to give to you from Jetbrains but Android Studio currently thinks that all Gradle scripts are Java 8. Last time I tried to bump the version of my scripts I got some funny errors about classpaths and Kotlin complaining that it couldn't inline a function whose bytecode is higher than the current bytecode (since it thinks the user's scripts are all Java 8 bytecodes).

You could however, downgrade Gradle to `7.X` again, and all this complications would disappear.

---

Actually, in my current project, I'm stuck with Gradle `8.0.2`, due to this issue: https://github.com/gradle/gradle/issues/25164

I've faced nightmares like that in the past, which is why I personally don't update Gradle that much often, especially in production.